[Style] Convert the 'font-palette' property to strong style types

weinig opened the pull request:
#### 1d2d25211b121f13e05343b2f0eaf2afbae7af3b
<pre>
[Style] Convert the &apos;font-palette&apos; property to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=299455">https://bugs.webkit.org/show_bug.cgi?id=299455</a>

Reviewed by Darin Adler.

Converts the &apos;font-palette&apos; property to strong style types.

Rather than duplicating the state of `font-palette`, we wrap
the platform type `WebCore::FontPalette` in a strong style
type `Style::FontPalette` which has all the style level information
on it. We may have to change this in the future when we add
support for palette-mix(), as we likely don&apos;t want to store
that on the FontCascadeDescriptor itself.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/CSSPrimitiveValue.h:
    - Add accessor for custom identifiers that asserts isCustomIdent().

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
    - Add a new codegen property for indicating that a font property
      wants to use getter and initial from RenderStyle rather than
      FontCascadeDescription. This will be removed when all font properties
      work this way, but allows doing that work incrementally.

* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialFontPalette): Deleted.
    - Moved initialFontPalette to RenderStyle, where it semantically makes sense.
      FontCascadeDescription really shouldn&apos;t know about initial values as that
      is a style system concept.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
    - Pipe the wrapper around.

* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
    - Remove old builder / extractor functions.

* Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
    - Add CSSPrimitiveValue overloads for the String, AtomString and CustomIdentifier
      converters to avoid some unnecessary checks when the caller already has one.

* Source/WebCore/style/values/fonts/StyleFontPalette.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontPalette.h: Added.
    - Adds new type with converter.

Canonical link: <a href="https://commits.webkit.org/300599@main">https://commits.webkit.org/300599@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b23381ab922d7c748a0048cb3a911461f16df913

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123215 "1 style error") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/42929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33626 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129922 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125092 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43653 "Built successfully") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51524 "Failed to checkout and rebase branch from PR 51269") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126168 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/34791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/110265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74295 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/122571 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/28421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73438 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104504 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28647 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132646 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/50165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/51524 "Failed to checkout and rebase branch from PR 51269") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/50541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/106486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/102019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25929 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/25588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/50020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/55781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49490 "Build is in progress. Recent messages:") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/52841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8466c163eb59a26a5d42fcdfc30ac881bdd4a92e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51269

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3cf71f43bdc081fc6b1997bb4989a6acfaa29848. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51269

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b23381ab922d7c748a0048cb3a911461f16df913. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51269

weinig committed:
[Style] Convert the 'font-palette' property to strong style types
https://bugs.webkit.org/show_bug.cgi?id=299455

Reviewed by Darin Adler.

Converts the 'font-palette' property to strong style types.

Rather than duplicating the state of `font-palette`, we wrap
the platform type `WebCore::FontPalette` in a strong style
type `Style::FontPalette` which has all the style level information
on it. We may have to change this in the future when we add
support for palette-mix(), as we likely don't want to store
that on the FontCascadeDescriptor itself.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/CSSPrimitiveValue.h:
    - Add accessor for custom identifiers that asserts isCustomIdent().

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
    - Add a new codegen property for indicating that a font property
      wants to use getter and initial from RenderStyle rather than
      FontCascadeDescription. This will be removed when all font properties
      work this way, but allows doing that work incrementally.

* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialFontPalette): Deleted.
    - Moved initialFontPalette to RenderStyle, where it semantically makes sense.
      FontCascadeDescription really shouldn't know about initial values as that
      is a style system concept.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
    - Pipe the wrapper around.

* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
    - Remove old builder / extractor functions.

* Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
    - Add CSSPrimitiveValue overloads for the String, AtomString and CustomIdentifier
      converters to avoid some unnecessary checks when the caller already has one.

* Source/WebCore/style/values/fonts/StyleFontPalette.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontPalette.h: Added.
    - Adds new type with converter.

Canonical link: https://commits.webkit.org/300599@main

webkit-commit-queue commented:
Committed 300599@main (1d2d25211b12): <https://commits.webkit.org/300599@main>

Reviewed commits have been landed. Closing PR #51269 and removing active labels.
